x-pack/plugins/security_solution/server/lib/detection_engine/rule_response_actions/endpoint_response_action.ts

```diff
@@ -5,81 +5,149 @@
+const NOOP = () => {};
 
-export const endpointResponseAction = (
+export const endpointResponseAction = async (
   responseAction: RuleResponseEndpointAction,
   endpointAppContextService: EndpointAppContextService,
-  { alerts }: ResponseActionAlerts,
-  experimentalFeatures: ExperimentalFeatures
-) => {
+  { alerts }: ResponseActionAlerts
+): Promise<void> => {
+  const logger = endpointAppContextService.createLogger(
+    'ruleExecution',
+    'automatedResponseActions'
+  );
+  const ruleId = alerts[0][ALERT_RULE_UUID];
+  const ruleName = alerts[0][ALERT_RULE_NAME];
+  const logMsgPrefix = `Rule [${ruleName}][${ruleId}]:`;
   const { comment, command } = responseAction.params;
+  const agentIds = uniq(map(alerts, 'agent.id'));
+  const errors: string[] = [];
+  const responseActionsClient = endpointAppContextService.getInternalResponseActionsClient({
+    agentType: 'endpoint',
+    username: 'unknown',
+  });
+
+  const automatedProcessActionsEnabled =
+    endpointAppContextService.experimentalFeatures.automatedProcessActionsEnabled;
+
+  const processResponseActionClientError = (err: Error, endpointIds: string[]): Promise<void> => {
+    errors.push(
+      `attempt to isolate host [${endpointIds.join(', ')}] returned error: ${err.message}`
+    );
 
-  const commonData = {
-    comment,
-    command,
-    rule_id: alerts[0][ALERT_RULE_UUID],
-    rule_name: alerts[0][ALERT_RULE_NAME],
-    agent_type: 'endpoint' as const,
+    return Promise.resolve();
   };
 
-  if (isIsolateAction(responseAction.params)) {
-    const alertsPerAgent = getIsolateAlerts(alerts);
-    each(alertsPerAgent, (actionPayload) => {
-      return endpointAppContextService.getActionCreateService().createActionFromAlert(
-        {
-          ...actionPayload,
-          ...commonData,
-        },
-        actionPayload.endpoint_ids
-      );
-    });
-  }
+  logger.info(`${logMsgPrefix} will create a total of [${agentIds.length}] [${command}] action(s)`);
+
+  let response: Promise<void> = Promise.resolve();
+
+  switch (command) {
+    case 'isolate':
+      response = Promise.all(
+        // TODO:PT should this be throttled?
+        Object.values(getIsolateAlerts(alerts)).map(
+          // eslint-disable-next-line @typescript-eslint/naming-convention
+          ({ endpoint_ids, alert_ids, parameters, error, hosts }: AlertsAction) => {
+            return responseActionsClient
+              .isolate(
+                {
+                  endpoint_ids,
+                  alert_ids,
+                  parameters,
+                  comment,
+                },
+                {
+                  hosts,
+                  ruleName,
+                  ruleId,
+                  error,
+                }
+              )
+              .catch((err) => {
+                return processResponseActionClientError(err, endpoint_ids);
+              })
+              .then(NOOP);
+          }
+        )
+      ).then(NOOP);
```

Comment:
why do we need NOOP ?

Reply:
to ensure it meets the outer function's expected return value of `Promise<void>`